feat: add OS version major/minor integer constants

jquick-axway opened the pull request:
**JIRA:**
https://jira.appcelerator.org/browse/TIMOB-28061

**Summary:**
Added new integer constants...
- `Ti.Platform.versionMajor`
- `Ti.Platform.versionMinor`
- `OS_VERSION_MAJOR`
- `OS_VERSION_MINOR`

**Reason:**
Makes it easier to implement new OS features. For example...
```javascript
if (OS_IOS && (OS_VERSION_MAJOR >= 13)) {
	// Do something on iOS 13 or higher only.
}
```

---
**Test:**
_Run this test on different Android/iOS versions._
1. Build and run the below on Android or iOS.
2. Verify `versionMajor` and `versionMinor` match `Ti.Platform.version` string.
3. Verify `OS_VERSION_MAJOR` and `OS_VERSION_MINOR` match `Ti.Platform.version` string.

```javascript
var window = Ti.UI.createWindow({ layout: "vertical" });
window.add(Ti.UI.createView({ height: OS_IOS ? "20dp" : "10dp" }));
window.add(Ti.UI.createLabel({
	text: "Ti.Platform.version: " + Ti.Platform.version,
	top: "10dp",
}));
window.add(Ti.UI.createLabel({
	text: "Ti.Platform.versionMajor: " + Ti.Platform.versionMajor,
	top: "10dp",
}));
window.add(Ti.UI.createLabel({
	text: "Ti.Platform.versionMinor: " + Ti.Platform.versionMinor,
	top: "10dp",
}));
window.add(Ti.UI.createLabel({
	text: "OS_VERSION_MAJOR: " + OS_VERSION_MAJOR,
	top: "10dp",
}));
window.add(Ti.UI.createLabel({
	text: "OS_VERSION_MINOR: " + OS_VERSION_MINOR,
	top: "10dp",
}));
window.open();
```

jquick-axway committed:
feat: add os version major/minor int constants

- Added new integer constants:
  * Ti.Platform.versionMajor
  * Ti.Platform.versionMinor
  * OS_VERSION_MAJOR
  * OS_VERSION_MINOR

Fixes TIMOB-28061

build commented:
<!--
  1 failure:  Tests have failed...
  1 warning:  tests/Resources/t...
  3 messages
  1 markdown notices
  DangerID: danger-id-Danger;
-->

<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>Tests have failed, see below for more information.</td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  **tests/Resources/ti.network.httpclient.test.js#L711** - tests/Resources/ti.network.httpclient.test.js line 711 – 'e' is defined but never used. Allowed unused args must match /^_.+/u. (no-unused-vars)
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>

  :floppy_disk: <a href="https://jenkins.appcelerator.org/job/titanium-sdk/job/titanium_mobile/job/PR-11859/1/artifact/dist/mobilesdk-9.2.0.v20200731222542-osx.zip">Here's the generated SDK zipfile</a>.
  </td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>:fist: The commits in this PR match our conventions! Feel free to Rebase and Merge this PR when ready.</td>
    </tr>
  
<tr>
      <td>:book:</td>
      <td>:x: 2 tests have failed
There are 2 tests failing and 708 skipped out of 7982 total tests.</td>
    </tr>
  </tbody>
</table>

### Tests:

<table><tr><th>Classname</th><th>Name</th><th>Time</th><th>Error</th></tr>
<tr><td>android.emulator.Titanium.UI.ImageView</td><td>image error event (9)</td><td>5.001</td><td><pre>Error: timeout of 5000ms exceeded
at Titanium.&lt;anonymous&gt; (/ti-mocha.js:6535:53723)</pre></td></tr>
<tr><td>ios.iphone.fs</td><td>#copyFile() copies file asynchronously to destination (13.3)</td><td>5.186</td><td><pre>Error: timeout of 5000ms exceeded
file:///Users/build/Library/Developer/CoreSimulator/Devices/F81C2DEA-4F67-4495-91CB-EBDEC27E4DFF/data/Containers/Bundle/Application/B9811A21-FD80-407C-96FA-07AC9FF502E2/mocha.app/ti-mocha.js:4290:32</pre></td></tr>
</table>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against d30e88bc7f0f698abc3bbde918cf29dbe19715d1
</p>

sgtcoolguy commented on apidoc/Titanium/Platform/Platform.yml:
```diff
@@ -357,10 +357,23 @@ properties:
+    description: Will return zero if OS does not have a minor version.
+    type: Number
+    permission: read-only
+    since: "9.2.0"
```
``` yml
default: `0`
```

Note that we haven't really tightened the definition for what is a valid "default" value. So it's really just for doc consumption on the site - but ideally we'd be able to more tightly specify this in some way that we could auto-generate tests or something from it.

jquick-axway replied:
Good idea. I'll do the same for `OS_VERSION_MINOR`.

sgtcoolguy commented on iphone/Resources/app.js:
```diff
@@ -20,3 +20,6 @@ btn.addEventListener('click', function() {
 win.add(btn);
 win.open();
+
+console.log("@@@ OS_VERSION_MAJOR: " + OS_VERSION_MAJOR);
```
Need to revert these changes

sgtcoolguy commented on apidoc/Global/Global.yml:
```diff
@@ -320,6 +320,25 @@ properties:
+
+  - name: OS_VERSION_MINOR
+    summary: The operating system's minor version number.
+    description: Will return zero if OS does not have a minor version.
```
link/refer to `Ti.Platform.versionMinor`?

sgtcoolguy commented on apidoc/Global/Global.yml:
```diff
@@ -320,6 +320,25 @@ properties:
     since: "9.0.0"
 
+  - name: OS_VERSION_MAJOR
+    summary: The operation system's major version number.
```
link/refer to `Ti.Platform.versionMajor`?

sgtcoolguy reviewed (changes requested):
This is awesome! It really shows how we're making it easier to add features cross-platform more easily too - the global defines in JS, the simplified JSC API on iOS, the already simple way to declare the property on Android.

My feedback is all nitpicks and none of it is required **except** the rollback of the change to the iOS app.js used by Xcode.

jquick-axway committed:
chore: updated TIMOB-28061 based on feedback

jquick-axway commented:
Updated PR:
- Updated API docs.
- Updated unit tests to verify `versionMajor`/`versionMinor` match components in `version` string.
- Removed test log message from iOS test app. _(My bad.)_

jquick-axway committed:
Merge branch 'master' into TIMOB-28061

jquick-axway commented:
@sgtcoolguy, what do you think of us adding our `OS_ANDROID` and `OS_IOS` constants to the new "globals.js" file?
```javascript
Object.defineProperties(global, {
	'OS_ANDROID': { value: OS_ANDROID, writable: false },
	'OS_IOS': { value: OS_IOS, writable: false },
	OS_VERSION_MAJOR: { value: Ti.Platform.versionMajor, writable: false },
	OS_VERSION_MINOR: { value: Ti.Platform.versionMinor, writable: false }
});
```

The reason to do this is because these constants are **undefined** if we're **NOT** running babel on the app project. This can happen:
- In our `titanium_mobile/android/app` test project under Android Studio.
- In our `titanium_mobile/iphone/Resources` test project under Xcode.
- In a Titanium app project if "tiapp.xml" has `<transpile>false</transplie>` set.

sgtcoolguy commented:
> @sgtcoolguy, what do you think of us adding our `OS_ANDROID` and `OS_IOS` constants to the new "globals.js" file?
> 
> ```js
> Object.defineProperties(global, {
> 	'OS_ANDROID': { value: OS_ANDROID, writable: false },
> 	'OS_IOS': { value: OS_IOS, writable: false },
> 	OS_VERSION_MAJOR: { value: Ti.Platform.versionMajor, writable: false },
> 	OS_VERSION_MINOR: { value: Ti.Platform.versionMinor, writable: false }
> });
> ```

Yeah, I'm totally fine with that. Makes sense to me.

jquick-axway committed:
feat: add OS_ANDROID/OS_IOS for non-transpiled builds

sgtcoolguy commented:
This feels very much like a PR that we should be able to merge without a formal QE FR, since we have unit tests to cover usage.

sgtcoolguy committed:
Merge branch 'master' into TIMOB-28061

jquick-axway commented:
@sgtcoolguy , I agree.
